### PR TITLE
Use validator tree of expressions and enforce no un-named properties.

### PR DESCRIPTION
**Pull request by mckoss**
Instead of merging all validation sub-expressions into one top-level expression, we now compute a "validator" - a tree of expression fragments that are AND-ed together when the final expressions are built.

This allows us to add a "wild-child" property with validate == false to prevent adding additional properties, as well as better merge disparate validation expressions.

We also make "type" a first-class AST value in preparation for allowing generic types (Map<String, Any>) for $key-validation expressions in derived versions of String.

### CODE REVIEW

**Comment by mckoss** [comment]
@TristonianJones Do you have time to take a look?

File: lib/rules-generator.js [hunk 1]
```diff
@@ -153,21 +162,28 @@ util.methods(Generator, {
...
     function registerAsCall(name, methodName) {
-      self.symbols.registerSchema(name, 'Any', undefined, {
+      self.symbols.registerSchema(name, ast.typeType('Any'), undefined, {
         validate: ast.method(['this'], ast.call(ast.reference(ast.cast(thisVar, 'Any'),
                                                               methodName)))
       });
     }
 
-    this.symbols.registerSchema('Any', 'Any', undefined, {
+    this.symbols.registerSchema('Any', ast.typeType('Any'), undefined, {
```
Comment (TristonianJones) [inline lib/rules-generator.js]: When referencing types here would it make sense to have the first two arguments be `Type` values? What you have seems functional, but maybe feels a little off-balance. The prior version of this method used `String` values for the first two args.

File: lib/rules-generator.js [hunk 2]
```diff
@@ -187,93 +203,102 @@ util.methods(Generator, {
...
-    for (var propName in schema.properties) {
-      var property = schema.properties[propName];
-      var propTerms = [];
-      for (var i = 0; i < property.types.length; i++) {
-        var type = property.types[i];
-        // @validator@<type>(this[propName])
-        propTerms.push(ast.call(ast.variable('@validator@' + type),
-                                [ ast.reference(thisVar, propName) ]));
+    var requiredProperties = [];
+    Object.keys(schema.properties).forEach(function(propName) {
```
Comment (TristonianJones) [inline lib/rules-generator.js]: I think there's a second argument for the object based forEach is the value, e.g. `function(propName, propType)`

Comment (mckoss) [inline lib/rules-generator.js]: IFAIK, forEach is only available for Array, not Object.  So using Object.keys() to get the keys as an array, and then enumerator over those.  It IS kind of gross.  I could add a objectForEach() type enumerator either from an npm package (overkill) or just add to util.js locally here.

File: lib/rules-generator.js (hunk 1 above)
Comment (mckoss) [inline lib/rules-generator.js]: In this case - we're registering a type in a symbol table (by name) - so it can be referenced by other type and path statements.  The "derivedFrom" has migrated to a TypeExpression - something that does not have to be a registered type, but can be one of UnionType, GenericType.

File: lib/rules-generator.js (hunk 2 above)
Comment (TristonianJones) [inline lib/rules-generator.js]: Sorry I was thinking of the new Map object:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/forEach

File: lib/rules-generator.js (hunk 1 above)
Comment (TristonianJones) [inline lib/rules-generator.js]: Thanks for clarifying

File: src/rules-parser.pegjs [hunk 3]
```diff
@@ -188,23 +187,43 @@ ParameterList = "(" _ ")" _ {
...
+  }
+  for (var i = 0; i < tail.length; i++) {
+    result.push(tail[i]);
+  }
+  return ast.unionType(result);
+}
+
+// Type, Type[], or Type<X, ... >
+// Type[] === Map<String, Type>
+SingleType = type:Identifier opt:(isMap:("\[\]") / "<" _ types:TypeList _ ">")? _ {
```
Comment (TristonianJones) [inline src/rules-parser.pegjs]: This change seems orthogonal to the validator expression trees, does this belong in another PR?

Comment (mckoss) [inline src/rules-parser.pegjs]: I guess this PR has:

- AST-based type definitions.
- Changing to validators in lieu of a single validation expression.

I think what you're proposing is not introducing new Type syntax in the parser in the PR (but retain the internal typeType?  I think that's a good suggestion.

Comment (TristonianJones) [inline src/rules-parser.pegjs]: Exactly, you read my mind. Retain the typeType but shift out the syntax changes to a different PR. Thanks!

Comment (mckoss) [inline src/rules-parser.pegjs]: Done.
